### PR TITLE
Add Cursor to ReadableBufferReader and dispatch from Pipe

### PR DESCRIPTION
**Pull request by pakrym**
@davidfowl

### CODE REVIEW

**Comment by davidfowl** [comment]
@pakrym why is the CI failing?

**Comment by pakrym** [comment]
@davidfowl because dispatched continuation was not awaited, removed dispatch from this PR.

File: src/System.IO.Pipelines/WritableBuffer.cs [hunk 1]
```diff
@@ -109,5 +109,14 @@ public Task FlushAsync()
         {
             return _pipe.FlushAsync();
         }
+
+        /// <summary>
+        /// Signals the <see cref="IPipelineReader"/> data is available.
+        /// </summary>
+        /// <returns>A task that completes when the data is fully flushed.</returns>
+        public Task SignalAsync()
```
Comment (davidfowl) [inline src/System.IO.Pipelines/WritableBuffer.cs]: I'm not a fan of this. What does it look like to use it?

File: src/System.IO.Pipelines/ReadableBuffer.cs [hunk 2]
```diff
@@ -223,7 +227,12 @@ public bool TrySliceTo(byte b1, out ReadableBuffer slice, out ReadCursor cursor)
         public ReadableBuffer Slice(int start, int length)
         {
             var begin = _start.Seek(start);
-            return Slice(begin, begin.Seek(length));
+            var end = begin.Seek(length);
```
Comment (davidfowl) [inline src/System.IO.Pipelines/ReadableBuffer.cs]: Why not bounds check as part of seek? Why do it after seeking?

**Comment by aL3891** [comment]
`public ReadCursor Move(ReadCursor cursor, int count)`, yes please :)

File: src/System.IO.Pipelines/WritableBuffer.cs (hunk 1 above)
Comment (pakrym) [inline src/System.IO.Pipelines/WritableBuffer.cs]: https://github.com/aspnet/KestrelHttpServer/commit/b1342aa4a5ab1975590a7179ac4d96ce22af6ac3#diff-4837ca204b80ef23b080588614cea7f8L302

File: src/System.IO.Pipelines/ReadableBuffer.cs (hunk 2 above)
Comment (pakrym) [inline src/System.IO.Pipelines/ReadableBuffer.cs]: There is a check in seek to make sure we are not Seeking past the end of segments, but we need also to check that we are not past the `End`. I was thinking about passing End to Seek and handling everything at one place.

File: src/System.IO.Pipelines/PipelineConnectionExtensions.cs [hunk 3]
```diff
@@ -49,7 +49,7 @@ public static ValueTask<int> ReadAsync(this IPipelineReader input, Span<byte> de
                 var inputBuffer = result.Buffer;
 
                 var fin = result.IsCompleted;
-                var sliced = inputBuffer.Slice(0, destination.Length);
+                var sliced = inputBuffer.Slice(0, Math.Min(inputBuffer.Length, destination.Length));
```
Comment (pakrym) [inline src/System.IO.Pipelines/PipelineConnectionExtensions.cs]: Yes, `Slice` lost it's "Slice at most" behavior and became "Slice exactly"

File: src/System.IO.Pipelines/ReadableBuffer.cs (hunk 2 above)
Comment (davidfowl) [inline src/System.IO.Pipelines/ReadableBuffer.cs]: Yes, pass end to Seek

File: src/System.IO.Pipelines/ReadableBuffer.cs [hunk 4]
```diff
@@ -551,5 +577,26 @@ bool ISequence<ReadOnlyMemory<byte>>.TryGet(ref Position position, out ReadOnlyM
             }
             return true;
         }
+
+        public ReadCursor Move(ReadCursor cursor, int count)
+        {
+            var newCursor = cursor.Seek(count);
```
Comment (davidfowl) [inline src/System.IO.Pipelines/ReadableBuffer.cs]: Also rename to Advance ala @aL3891 's comment

Comment (pakrym) [inline src/System.IO.Pipelines/ReadableBuffer.cs]: Advance is to similar to AdvanceReader

**Pull request merged**